source/common/access_log/access_log_formatter.cc

```diff
@@ -64,14 +66,61 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
   std::string log_line;
   log_line.reserve(256);
 
-  for (const FormatterPtr& formatter : formatters_) {
-    log_line +=
-        formatter->format(request_headers, response_headers, response_trailers, stream_info);
+  for (const FormatterProviderPtr& provider : providers_) {
+    log_line += provider->format(request_headers, response_headers, response_trailers, stream_info);
   }
 
   return log_line;
 }
 
+JsonFormatterImpl::JsonFormatterImpl(
+    const std::map<const std::string, const std::string>& format_mapping) {
+  for (const auto& pair : format_mapping) {
+    auto providers = AccessLogFormatParser::parse(pair.second);
+
+    // Enforce that each key only has one format specifier in it
+    if (providers.size() > 1) {
+      throw EnvoyException(fmt::format(
+          "More than one format specifier was provided in the JSON log format: {}", pair.second));
+    }
+
+    if (providers.size() < 1) {
+      throw EnvoyException(
+          fmt::format("No format specifier was provided in the JSON log format: {}", pair.second));
+    }
+
+    json_output_format_.emplace(pair.first, std::move(providers[0]));
+  }
+}
+
+std::string JsonFormatterImpl::format(const Http::HeaderMap& request_headers,
+                                      const Http::HeaderMap& response_headers,
+                                      const Http::HeaderMap& response_trailers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  rapidjson::StringBuffer strbuf;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(strbuf);
+
+  auto output_map = this->to_map(request_headers, response_headers, response_trailers, stream_info);
+  writer.StartObject();
+  for (const auto& pair : output_map) {
+    writer.Key(pair.first.c_str());
+    writer.String(pair.second.c_str());
+  }
+  writer.EndObject();
+  return fmt::format("{}\n", strbuf.GetString());
```

Comment:
The alternative to using RapidJSON here is to build a `google.protobuf.Struct` and then do a Proto3 --> JSON conversion. I kind of prefer this, unless you want fine grained control over the output format.

Reply:
I can look into this. Shouldn't be too complex, although there is a fair bit of boilerplate with constructing Proto3 objects (see the unit tests I wrote in `config.cc`). But I'll move forward with this anyway.

- [ ] use protobuf to serialize instead of JSON

Reply:
Turned out to be pretty straightforward